enhancements/network/user-space-pod-iface-and-api-library.md

```diff
@@ -0,0 +1,262 @@
+status: implementable
+---
+
+# User Space Pod Interface and API Library
+
+Add OCP support for user space applications (i.e. - DPDK based applications)
+running in pods.
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+A typical pod has one and only one kernel interface (one side of a veth-pair).
+Interface attributes such as IP address and MAC address are applied to the
+interface through the kernel. More advance use-cases involve injecting more than
+one interface into the pod using meta-CNI plugins such as Multus. In this case,
+there are several scenarios that require additional information passed to the
+pod:
+- Multiple interfaces:
+  - With multiple interfaces now in the pod, the pod needs context on which
+  interface is intended for what purpose.
+  - Needed in most multi-interface scenarios, but specifically OCP related,
+  when multiple SR-IOV VFs are injected into a pod.
+- Kernel interface is used by user space application in the pod:
+  - When the container workload performs packet processing in user space,
+  for example a DPDK application, the interface is no longer managed by the
+  kernel in the pod, and data like IP address and MAC address can not use the
+  kernel to pass this information.
+  - In addition to data typically passed via the kernel interface, user space
+  applications require additional information to function properly. Examples
+  of some of the data include:
+    - CPUs allocated to the pod
+    - Number of hugepages dedicated to the pod
+    - Number of queues used by the interface
+  - Needed in most user space application scenarios, but specifically OCP
+  related, when user space applications use SR-IOV VFs in a pod.
+- User space interface is injected into the pod:
+  - A user space interface, like a pure user space interface such as
+  vhost-user (not supported by OCP) or a VDPA interface, which uses the
+  virtio protocol for data plane negotiation (support expect in future OCP
+  release, around 4.8 or 4.9 time frame), requires additional information like
+  Unix socket file and virtio mode.
+  - Not needed by OCP immediately, but must be considered during the overall
+  design.
+
+This enhancement has two components:
+- How the data should be passed to the pod?
+- How to ease the consumption of the data in the pod?
+
+## Motivation
+
+Various system resources are allocated to run high performance network
+applications inside a container such as CPU cores, dedicated memory (hugepages),
+network devices, etc. However, these resources are presented differently inside
+the container and some of them are not visible to applications without
+explicitly injecting to the container.
+
+### Goals
```

Comment:
I guess the template doesn't say this, but this section is usually done as fairly precise bullet points. eg

- Define how data is passed from Kubernetes to user space networking applications
- Provide a library to be used for developers to blah blah blah
- Ensure customers using SR-IOV can something something something
- etc

Reply:
Done